### PR TITLE
Adding gzip compression (dynamic gzip filter)

### PR DESCRIPTION
**Pull request by Antwnis**
Adding optional configuration `gzip` to enable dynamic gzip compression

In our environment with gzip we noticed network I/O dropping to 1 / 14th 
i.e. 176,270 byte / sec --> 12,801 byte / sec rate

### CODE REVIEW

File: jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java [hunk 1]
```diff
@@ -18,6 +19,9 @@ public static void main(String[] args) throws Exception {
      Server server = new Server(port);
      ServletContextHandler context = new ServletContextHandler();
      context.setContextPath("/");
+     if (jc.gzipEnabled()) {
```
Comment (brian-brazil) [inline jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java]: Just hardcode it to on

File: jmx_prometheus_httpserver/pom.xml [hunk 2]
```diff
@@ -30,6 +30,11 @@
       <artifactId>jetty-servlet</artifactId>
       <version>8.1.7.v20120910</version>
     </dependency>
+    <dependency>
+      <groupId>org.eclipse.jetty</groupId>
+      <artifactId>jetty-servlets</artifactId>
+      <version>8.1.7.v20120910</version>
```
Comment (brian-brazil) [inline jmx_prometheus_httpserver/pom.xml]: While you're here, would you mind upgrading us to something recent to solve #94?

File: jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java [hunk 3]
```diff
@@ -5,6 +5,7 @@
 import org.eclipse.jetty.server.Server;
 import org.eclipse.jetty.servlet.ServletContextHandler;
 import org.eclipse.jetty.servlet.ServletHolder;
+import org.eclipse.jetty.servlets.GzipFilter;
 
 public class WebServer {
```
Comment (brian-brazil) [inline jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java]: The java agent (which is the preferred way of doing things) also need changing.

File: jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java (hunk 1 above)
Comment (Antwnis) [inline jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java]: not sure if we should.. i'd say keep it optional - and present it as a recommended setting on the docs.. WDYT?

File: jmx_prometheus_httpserver/pom.xml (hunk 2 above)
Comment (Antwnis) [inline jmx_prometheus_httpserver/pom.xml]: Shall we introduce 0.0.18 version of the io.prometheus dependencies and make a 0.0.18 release of the jmx_exporter ?

**Comment by Antwnis** [comment]
@brian-brazil we are all too busy - but what do u think of the above ?! 
Are you happy with all parts | some parts of this ?!

File: jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java (hunk 1 above)
Comment (brian-brazil) [inline jmx_prometheus_httpserver/src/main/java/io/prometheus/jmx/WebServer.java]: If someone wants to turn it off we can consider it, but I'd avoid a knob we don't absolutely have to have.

File: jmx_prometheus_javaagent/src/main/java/io/prometheus/jmx/JavaAgent.java [hunk 4]
```diff
@@ -34,15 +35,20 @@ public static void premain(String agentArgument, Instrumentation instrumentation
        file = args[1];
      }
 
-     new JmxCollector(new File(file)).register();
+     JmxCollector jc = new JmxCollector(new File(file)).register();
      DefaultExports.initialize();
 
      server = new Server(socket);
+     /*
```
Comment (brian-brazil) [inline jmx_prometheus_javaagent/src/main/java/io/prometheus/jmx/JavaAgent.java]: Debug code?

**Comment by Antwnis** [comment]
Please review - two commits : 
1. Add gzip to the java agent
 
https://github.com/prometheus/jmx_exporter/pull/95/commits/5235373d7dfe983cd4f5eb3b252b0e277794089b

2. Trying to fix #94 

https://github.com/prometheus/jmx_exporter/pull/95/commits/366d0a8aed8e7ff3ab1b9f6bf265bf78504fb891

On that second one - after upgrading to more recent Jetty - we can no longer explicitly set both the QueuedPool AND a Socket - on new version we can set JUST one of them 

So i'm not too sure if this has any side effects
We need to set either of them - and socket takes preference

**Comment by mzagar** [comment]
hi all, any updates on this PR and when it could be merged?
we really need compression - anyway we can help?

**Comment by brian-brazil** [comment]
We've just switched away from jetty so this PR no longer applies, however if you'd like to add transparent gzip support over in client_java's simpleclient_httpserver that'd be welcome.

**Pull request closed without merging**